components/infobox/wikis/starcraft/infobox_person_player_custom.lua

```diff
@@ -0,0 +1,510 @@
+local Cell = require('Module:Infobox/Widget/Cell')
+
+local _CURRENT_YEAR = tonumber(os.date('%Y'))
+local _ALLOWED_PLACES = {'1', '2', '3', '4', '3-4'}
+local _EARNING_MODES = {['solo'] = '1v1', ['team'] = 'team'}
+local _MAXIMUM_NUMBER_OF_PLAYERS_IN_PLACEMENTS = 30
+local _NUMBER_OF_ALLOWED_ACHIEVEMENTS = 10
+local _FIRST_DAY_OF_YEAR = '-01-01'
+local _LAST_DAY_OF_YEAR = '-12-31'
+local _KOREAN = 'South Korea'
+
+--race stuff
+--local _AVAILABLE_RACES = {'p', 't', 'z', 'r', 'total'}
+local _RACE_FIELD_AS_CATEGORY_LINK = true
+local _RACE_DATA = {
+	p = {'Protoss'},
+	pt = {'Protoss', 'Terran'},
+	pz = {'Protoss', 'Zerg'},
+	t = {'Terran'},
+	tp = {'Terran', 'Protoss'},
+	tz = {'Terran', 'Zerg'},
+	z = {'Zerg'},
+	zt = {'Zerg', 'Terran'},
+	zp = {'Zerg', 'Protoss'},
+	r = {'Random'},
+	a = {'Protoss', 'Terran', 'Zerg'},
+}
+local _RACE_ALL = 'All'
+local _RACE_ALL_SHORT = 'a'
+
+local _earningsGlobal = {}
+local _achievements = {}
+local _awardAchievements = {}
+local _raceData
+local _statusStore
+
+local CustomPlayer = Class.new()
+
+local CustomInjector = Class.new(Injector)
+
+local _args
+local _player
+local _lpdbData = {}
+
+function CustomPlayer.run(frame)
+	local player = Person(frame)
+	_args = player.args
+	_player = player
+
+	player.getStatusToStore = CustomPlayer.getStatusToStore
+	player.adjustLPDB = CustomPlayer.adjustLPDB
+	player.nameDisplay = CustomPlayer.nameDisplay
+	player.calculateEarnings = CustomPlayer.calculateEarnings
+	player.createWidgetInjector = CustomPlayer.createWidgetInjector
+	player.getWikiCategories = CustomPlayer.getWikiCategories
+	player.getPersonType = CustomPlayer.getPersonType
+
+	return player:createInfobox(frame)
+end
+
+function CustomInjector:parse(id, widgets)
+	if id == 'status' then
+		return {
+			Cell{
+				name = 'Race',
+				content = {CustomPlayer._getRaceData(_args.race or 'unknown', _RACE_FIELD_AS_CATEGORY_LINK)}
+			}
+		}
+	elseif id == 'role' then return {}
+	elseif id == 'region' then return {}
+	elseif
+		id == 'history' and
+		string.match(_args.retired or '', '%d%d%d%d')
+	then
+		table.insert(widgets, Cell{
+				name = 'Retired',
+				content = {_args.retired}
+			})
+	end
+	return widgets
+end
+
+function CustomInjector:addCustomCells(widgets)
+	--enable this AFTER the match to match2 conversion of 1v1 matchlists and brackets
+	--local yearsActive = Namespace.isMain() and CustomPlayer._getMatchupData() or nil
+
+	local currentYearEarnings = _earningsGlobal[tostring(_CURRENT_YEAR)]
+	if currentYearEarnings then
+		currentYearEarnings = Math.round{currentYearEarnings}
+		currentYearEarnings = '$' .. mw.language.new('en'):formatNum(currentYearEarnings)
+	end
+
+	return {
+		Cell{
+			name = 'Approx. Winnings ' .. _CURRENT_YEAR,
+			content = {currentYearEarnings}
+		},
+		--Cell{name = 'Years active', content = {yearsActive}}
+	}
+end
+
+function CustomPlayer.nameDisplay()
+	CustomPlayer._getRaceData(_args.race or 'unknown')
```

Comment:
Why is "or unknown" needed?

Reply:
moved it to the function
(unknown (question mark) icon is to be displayed if no race is entered)